vllm/inputs/data.py

```diff
@@ -103,39 +44,41 @@ class TokensPrompt(TypedDict):
 which encapsulates multiple prompts, i.e. of the sort
 which may be utilized for encoder/decoder models when
 the user desires to express both the encoder & decoder
-prompts explicitly, i.e. ExplicitEncoderDecoderPrompt
+prompts explicitly, i.e. :class:`ExplicitEncoderDecoderPrompt`
 
-A prompt of type SingletonPromptInputs may be employed
+A prompt of type :class:`SingletonPromptInputs` may be employed
 as (1) input to a decoder-only model, (2) input to
 the encoder of an encoder/decoder model, in the scenario
 where the decoder-prompt is not specified explicitly, or
 (3) as a member of a larger data structure encapsulating
-more than one prompt, i.e. ExplicitEncoderDecoderPrompt
+more than one prompt, i.e. :class:`ExplicitEncoderDecoderPrompt`
 """
 
+_T = TypeVar("_T", bound=SingletonPromptInputs, default=SingletonPromptInputs)
+
 
-class ExplicitEncoderDecoderPrompt(TypedDict):
+class ExplicitEncoderDecoderPrompt(TypedDict, Generic[_T]):
     """Represents an encoder/decoder model input prompt,
     comprising an explicit encoder prompt and a 
     decoder prompt.
 
     The encoder and decoder prompts, respectively,
     may formatted according to any of the
-    SingletonPromptInputs schemas, and are not
+    :class:`SingletonPromptInputs` schemas, and are not
     required to have the same schema.
 
     Only the encoder prompt may have multi-modal data.
 
-    Note that an ExplicitEncoderDecoderPrompt may not
+    Note that an :class:`ExplicitEncoderDecoderPrompt` may not
     be used as an input to a decoder-only model,
     and that the `encoder_prompt` and `decoder_prompt`
-    fields of this data structure may not themselves
-    must be SingletonPromptInputs instances.
+    fields of this data structure themselves must be
+    :class:`SingletonPromptInputs` instances.
     """
 
-    encoder_prompt: SingletonPromptInputs
+    encoder_prompt: _T
 
-    decoder_prompt: SingletonPromptInputs
+    decoder_prompt: Optional[_T]
```

Comment:
Using generics makes sense. Correct me if I am wrong - since the current implementation has a single generic type `_T`, do `encoder_prompt` and `decoder_prompt` have to have the same type?

For example, would this prompt be compatible with `ExplicitEncoderDecoderPrompt`?:

```
{
'encoder_prompt': 'This is an encoder prompt', # raw text prompt
'decoder_prompt': [2] # token list prompt
}
```

The scenario I have in mind is that a vLLM request passes in a raw string encoder prompt (i.e. a question provided by a user) but also wants to explicitly specify a list of control tokens for the decoder prompt in order to affect model behavior.

```suggestion
_T1 = TypeVar("_T1", bound=SingletonPromptInputs, default=SingletonPromptInputs)
_T2 = TypeVar("_T2", bound=SingletonPromptInputs, default=SingletonPromptInputs)


class ExplicitEncoderDecoderPrompt(TypedDict, Generic[_T1,_T2]):
    """Represents an encoder/decoder model input prompt,
    comprising an explicit encoder prompt and a 
    decoder prompt.

    The encoder and decoder prompts, respectively,
    may formatted according to any of the
    :class:`SingletonPromptInputs` schemas, and are not
    required to have the same schema.

    Only the encoder prompt may have multi-modal data.

    Note that an :class:`ExplicitEncoderDecoderPrompt` may not
    be used as an input to a decoder-only model,
    and that the `encoder_prompt` and `decoder_prompt`
    fields of this data structure themselves must be
    :class:`SingletonPromptInputs` instances.
    """

    encoder_prompt: _T1

    decoder_prompt: Optional[_T2]
```

Are there downsides to this approach?

Reply:
Hi @DarkLight1337 I see you already made this change, thanks. 

Just to circle back and answer your question - 

This portion of the encoder/decoder example code
https://github.com/vllm-project/vllm/pull/7258/files#diff-9f99d5748f1e063be23804194b63b2b0bb8c1e314d5a35453678e6cbd2996a3aR44-R49

tests a string encoder prompt & an explicit token decoder prompt. It is a supported prompt combination. I also highlighted where in the code this is supported
https://github.com/vllm-project/vllm/pull/7258#discussion_r1707318010

Ultimately when the encoder/decoder prompts reach the backend they are both token lists, so supporting different prompt types is no issue.